shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -122,38 +155,63 @@ private case class CloseableColumnBatchIterator(iter: Iterator[ColumnarBatch]) e
+  override def supportsColumnarOutput(schema: StructType): Boolean = schema.fields.forall(f =>
+    // only check spark b/c if we are on the GPU then we will be calling the gpu method regardless
+    isSupportedBySparkColumnar(f.dataType))
+
+  private def isSupportedBySparkColumnar(dataType: DataType): Boolean =
+    // AtomicType check. Everything is supported besides Structs, Maps or UDFs
+    dataType match {
+      case TimestampType | StringType | BooleanType | DateType | BinaryType |
+           DoubleType | FloatType | ByteType | IntegerType | LongType | ShortType => true
+      case _: HiveStringType => true
+      case _: DecimalType => true
+      case _ => false
+    }
+
+  def isSupportedByCudf(schema: Seq[Attribute]): Boolean = {
+    schema.forall(a => GpuColumnVector.isSupportedType(a.dataType))
+  }
 
   /**
    * Convert an `RDD[ColumnarBatch]` into an `RDD[CachedBatch]` in preparation for caching the data.
    * This method uses Parquet Writer on the GPU to write the cached batch
+   *
    * @param input the input `RDD` to be converted.
    * @param schema the schema of the data being stored.
    * @param storageLevel where the data will be stored.
    * @param conf the config for the query.
    * @return The data converted into a format more suitable for caching.
    */
   override def convertColumnarBatchToCachedBatch(input: RDD[ColumnarBatch],
-                                       schema: Seq[Attribute],
-                                       storageLevel: StorageLevel,
-                                       conf: SQLConf): RDD[CachedBatch] = {
-    def putOnGpuIfNeeded(batch: ColumnarBatch): ColumnarBatch = {
-      if (batch.numCols() > 0 && !batch.column(0).isInstanceOf[GpuColumnVector]) {
-        val s = StructType(
-          schema.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
-        val gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
-        batch.close()
-        gpuCB
-      } else {
-        batch
+     schema: Seq[Attribute],
+     storageLevel: StorageLevel,
+     conf: SQLConf): RDD[CachedBatch] = {
+    val rapidsConf = new RapidsConf(conf)
+    if (rapidsConf.isSqlEnabled && isSupportedByCudf(schema)) {
+      def putOnGpuIfNeeded(batch: ColumnarBatch): ColumnarBatch = {
+        if (!batch.column(0).isInstanceOf[GpuColumnVector]) {
+          val s: StructType = schema.toStructType
+          val gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
+          batch.close()
+          gpuCB
+        } else {
+          batch
+        }
       }
-    }
 
-    input.map(batch => {
-      withResource(putOnGpuIfNeeded(batch)) { gpuCB =>
-        compressColumnarBatchWithParquet(gpuCB)
-      }
-    })
+      input.map(batch => {
+        if (batch.numCols() == 0) {
+          ParquetCachedBatch(batch.numRows(), new Array[Byte](0))
+        }
+        withResource(putOnGpuIfNeeded(batch)) { gpuCB =>
```

Comment:
Won't line 205 return? I have added `else` for readability.

Reply:
Why would line 205 return?   It's just a statement _that is not the last statement_.  It will create a `ParquetCacheBatch` instance which is instantly unreferenced then proceeds to line 207.

If you want to return early then that's what the `return` keyword is for. :smile:

Reply:
:face_with_head_bandage: 
*last statement* is what I forgot from my limited knowledge of scala. Thanks